File: pomade-signer/src/mailer/resend.rs
use serde_json::json;

use crate::mailer::{Email, MailFuture, Mailer};

pub struct ResendMailer {
    pub client: reqwest::Client,
    pub api_key: String,
}

impl Mailer for ResendMailer {
    fn send(&self, from_email: &str, from_name: &str, email: Email) -> MailFuture {
        let client = self.client.clone();
        let key = self.api_key.clone();
        let from = format!("{} <{}>", from_name, from_email);
        Box::pin(async move {
            let res = client
                .post("https://api.resend.com/emails")
                .bearer_auth(&key)
                .json(&json!({
                    "from": from,
                    "to": [email.to],
                    "subject": email.subject,
                    "text": email.text,
                    "html": email.html,
                }))
                .send()
                .await
                .map_err(|e| e.to_string())?;

            if res.status().is_success() {
                Ok(())
            } else {
                Err(format!("resend error: {}", res.status()))
            }
        })
    }
}

File: pomade-signer-go/mailer/resend.go
package mailer

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

type ResendMailer struct {
	Client *http.Client
	APIKey string
}

func (m ResendMailer) Send(ctx context.Context, fromEmail string, fromName string, email Email) error {
	body, _ := json.Marshal(map[string]any{
		"from":    fmt.Sprintf("%s <%s>", fromName, fromEmail),
		"to":      []string{email.To},
		"subject": email.Subject,
		"text":    email.Text,
		"html":    email.HTML,
	})
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.resend.com/emails", bytes.NewReader(body))
	req.Header.Set("Authorization", "Bearer "+m.APIKey)
	req.Header.Set("Content-Type", "application/json")
	res, err := m.Client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode > 299 {
		return fmt.Errorf("resend error: %s", res.Status)
	}
	return nil
}
